apps/evm/README.md

```diff
@@ -1,43 +1,160 @@
+* `State:` The world state of Ethereum, defined as the root hash of a Merkle Patricia Trie containing all account data. See Section 4.1 of the Yellow Paper, or the [merkle_patricia_tree](../merkle_patricia_tree) umbrella project in this repo.
+* `Machine State:` This structure effectively encodes the current context of a running VM. The Machine State is used during execution of the program and discarded on completion. Before we finish, we extract the gas used and return the value from this object. The machine state includes:
+  * the program counter
+  * memory contents
+  * active number of words in memory
+  * the stack contents
+* `Sub State:` The sub state tracks the selfdestruct list (contracts to destroy), the log, and the refund (for cleaning up storage) for a contract execution.
+* `Execution Environment:` This tracks information about the call into a contract, such as the machine code itself and the value passed to the contract or message call. Other than stack depth, this is generally not mutated during execution of the machine.
 
-# Examples
+### File Structure Notes
+- All operations as defined in the yellow paper are located in the [lib/evm/operation](lib/evm/operation) folder.
+- [liv/evm/vm.ex](lib/evm/vm.ex) contains the following virtual machine functions from the yellow paper:
+  - `EVM.VM.run/2`: Ξ function.
+  - `EVM.VM.exec/3`: X function.
+  - `EVM.VM.cycle/3`: O function.
+  - `EVM.Functions.is_exception_halt?/2`: Z function.
 
-Here is an example of a simple program running on the VM:
 
-```elixir
-EVM.VM.run(%{}, 5, %EVM.ExecEnv{machine_code: EVM.MachineCode.compile([:push1, 3, :push1, 5, :add, :push1, 0x00, :mstore, :push1, 32, :push1, 0, :return])})
-{%{}, 5, [], [], 0, <<0x08::256>>}
+## Installation
+
+Installation is handled through the bin/setup procedure in the [Mana-Ethereum README](../../README.md).
+
+## Example
+
+The following example is from [Ethereum Virtual Machine in Elixir](https://www.badykov.com/elixir/2018/04/29/evm-basics/). It illustrates the `VMTests/vmArithmeticTest/add3.json` test from the [Ethereum common test protocol](https://github.com/ethereum/tests/).
+
+### Setup
+
+1. Clone the repo.  `git clone https://github.com/poanetwork/mana`
+
+2. Fetch dependencies.  `./bin/setup`
+
+3. Add code to print debugging information.
+   - Go to `lib/evm/vm.ex`.
+
+   - Find the `cycle/3` method and **add the IO code**.
+      ```elixir
+      def cycle(machine_state, sub_state, exec_env) do
+      operation = MachineCode.current_operation(machine_state, exec_env)
+      inputs = Operation.inputs(operation, machine_state)
+      
+      IO.puts("stack:")
+      IO.inspect(machine_state.stack)
+      IO.puts("operation: #{operation.sym}")
+      
+      ```
+   - Save the file.
+
+ 4. Start the Elixir REPL.  `cd apps/evm && iex -S mix`
```

Comment:
This is under the `setup` section, but it seems to have more do to with seeing how the EVM works. Should this be more of an example section?